physicalplan/function/function.go

```diff
@@ -1,101 +1,279 @@
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, true, true, f.StepTime, f.SelectRange)
 		}))
 	},
-	"delta": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreSingleValue(points, sample(labels, stepTime, points, func(points []promql.Point) float64 {
-			return extrapolatedRate(points, false, false, stepTime, selectRange)
+	"delta": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, false, false, f.StepTime, f.SelectRange)
 		}))
 	},
-	"increase": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreSingleValue(points, sample(labels, stepTime, points, func(points []promql.Point) float64 {
-			return extrapolatedRate(points, true, false, stepTime, selectRange)
+	"increase": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, true, false, f.StepTime, f.SelectRange)
 		}))
 	},
-	"resets": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreEmpty(points, sample(labels, stepTime, points, resets))
+	"resets": func(f FunctionArgs) promql.Sample {
+		return ignoreEmpty(f.Points, sample(f.Labels, f.StepTime, f.Points, resets))
+	},
+	"clamp": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clamp))
 	},
+	"clamp_min": func(f FunctionArgs) promql.Sample {
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMin))
+	},
+	"clamp_max": func(f FunctionArgs) promql.Sample {
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMax))
+	},
+}
+
+// functionSelector returns []model.StepVector after processing input with desired function.
+type functionSelector struct {
+	vectorPool    *model.VectorPool
+	funcExpr      *parser.Call
+	mint          int64
+	maxt          int64
+	step          int64
+	currentStep   int64
+	stepsBatch    int
+	series        []labels.Labels
+	once          sync.Once
+	majorArgIndex int
+	majorArg      model.VectorOperator
+	scalarArgs    []model.VectorOperator
+
+	call FunctionCall
+}
+
+func NewFunctionSelector(pool *model.VectorPool, funcExpr *parser.Call, mint, maxt time.Time, step time.Duration, stepsBatch int, call FunctionCall, majorArgIndex int, majorArg model.VectorOperator, scalarArgs ...model.VectorOperator) model.VectorOperator {
+	return &functionSelector{
+		vectorPool:    pool,
```

Comment:
I think this operator can make all changes in-place, like for the unary operator. So we should not need a new pool here, and we can return the `next` pool in the `GetPool()` function.